### PR TITLE
Handle nullability and nesting

### LINKED ISSUE #10: [Code Generation] Handle type nullability
Add with `@Nullable` annotations to methods

### PR DESCRIPTION
**Pull request by felipecsl**
this PR was supposed to handle just the Nullability part, but turned out
to be much bigger since once I started to implement that, I started
bumping into roadblocks with the current implementation that wouldn't
let me properly propragate the nullability information without
refactoring a lot of code.
It ended up being much easier to also make the change to generate
nested types with the correct names based on the responseName field.
The code turned out to be **at lot** simpler and much more similar to
the Swift counterparts. This still doesnt handle deduping of duplicate
types but that shouldnt be too hard to implement now. We just need to
decide on how to actualy do it (maybe add a suffix to the interface
name? Maybe a number, or a letter, or a symbol?) I'm open to ideas here.

Fixes #10 and #17

### CODE REVIEW

**Comment by martijnwalraven** [comment]
Sounds great! I agree with [your comment](https://github.com/apollostack/apollo-android/pull/15#discussion_r90968893) to only deduplicate when needed.

Something like `interface HeroFriendsFriends` makes most semantic sense, but feels weird when nested because that means the qualified name already has that same information. So maybe just a simple number suffix, like `Friends2`?

Is there any precedence for these types of workarounds in Java / Android? (I still don't understand the restriction, because it seems the type being nested and having a different fully qualified name should be enough to disambiguate.)

File: apollo-compiler/src/main/kotlin/com/apollostack/compiler/ir/GraphQlType.kt [hunk 1]
```diff
@@ -20,30 +22,47 @@ sealed class GraphQlType(val nullable: Boolean) {
...
-    }
-  }
+  fun toJavaTypeName() = graphQlTypeToJavaTypeName(this, !nullable, nullable)
 
   companion object {
-
-    private fun String.normalizeTypeName() = removeSuffix("!").removeSurrounding(prefix = "[", suffix = "]").removeSuffix("!")
+    private val NULLABLE_ANNOTATION = AnnotationSpec.builder(Nullable::class.java).build()
+    private val LIST_TYPE = ClassName.get(List::class.java)
+    private val GRAPHQLTYPE_TO_JAVA_TIME = mapOf(
```
Comment (sav007) [inline apollo-compiler/src/main/kotlin/com/apollostack/compiler/ir/GraphQlType.kt]: TIME you probably meant TYPE?

File: apollo-compiler/src/main/kotlin/com/apollostack/compiler/ir/Field.kt [hunk 2]
```diff
@@ -1,31 +1,46 @@
...
+        .addModifiers(Modifier.PUBLIC, Modifier.ABSTRACT)
+        .build()
   }
 
-  private fun String.toTypeName(): TypeName = GraphQlType.resolveByName(this).toJavaTypeName()
+  private fun toTypeName(responseType: String): TypeName =
+      GraphQlType.resolveByName(responseType).toJavaTypeName()
+
+  private fun methodResponsType(): String {
+    // TODO: Do we also need to handle nested lists?
```
Comment (sav007) [inline apollo-compiler/src/main/kotlin/com/apollostack/compiler/ir/Field.kt]: I thought our GraphQlType class supports nested lists, I remember I even added test case for that, or I missing smth?

File: apollo-compiler/src/main/kotlin/com/apollostack/compiler/OperationTypeSpecBuilder.kt [hunk 3]
```diff
@@ -1,58 +1,26 @@
...
+  private fun buildTypeSpecs(currentTypeName: String, fields: List<Field>): TypeSpec {
+    val innerTypes = fields
+        .filter(Field::isNonScalar)
+        .map { buildTypeSpecs(it.responseName.capitalize().singularize(), it.fields!!) }
+    return TypeSpec.interfaceBuilder(currentTypeName)
+        .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
+        .addMethods(fields.map(Field::toMethodSpec))
+        .addTypes(innerTypes)
+        .build()
   }
```
Comment (sav007) [inline apollo-compiler/src/main/kotlin/com/apollostack/compiler/OperationTypeSpecBuilder.kt]: wow, nice so simpler :)

**Comment by sav007** [comment]
As for dup nested declaration I guess we need to add prefix|suffix to the inner interface only, so as eaxample to fix this issue:
```
  static interface Hero {

    static interface Friend {

      static interface Node {

        static interface Friend {

        }
      }
    }
  }
```

we need to turn it into: 
```
static interface Hero {

    static interface Friend {

      static interface Node {

        static interface NodeFriend {

        }
      }
    }
  }
```

And that is all, we don't need to prefix|suffix them all. And the algorithm should be simple, we just track if we have already defined outer interface with the same name, we just add prefix - the name of surrounding. Or even simpler just add "_" :)

**Comment by sav007** [comment]
Examples:

```
static interface Hero {

    static interface Friend {

      static interface Node {

        static interface NodeFriend {
          
          static interface NodeFriendHero {
            
            static interface NodeFriendHeroFriend {
              
            } 
          }

        }
      }
    }
  }
```
or
```
static interface Hero {

    static interface Friend {

      static interface Node {

        static interface Friend1 {

          static interface Hero1 {

            static interface Friend2 {

            }
          }

        }
      }
    }
  }
```

File: apollo-compiler/src/main/kotlin/com/apollostack/compiler/ir/GraphQlType.kt [hunk 4]
```diff
@@ -20,30 +22,47 @@ sealed class GraphQlType(val nullable: Boolean) {
...
-      is GraphQLList -> ParameterizedTypeName.get(ClassName.get(List::class.java), listType.toJavaTypeName())
-      is GraphQlUnknown -> ClassName.get("", typeName)
-    }
-  }
+  fun toJavaTypeName() = graphQlTypeToJavaTypeName(this, !nullable, nullable)
 
   companion object {
-
-    private fun String.normalizeTypeName() = removeSuffix("!").removeSurrounding(prefix = "[", suffix = "]").removeSuffix("!")
+    private val NULLABLE_ANNOTATION = AnnotationSpec.builder(Nullable::class.java).build()
```
Comment (marwanad) [inline apollo-compiler/src/main/kotlin/com/apollostack/compiler/ir/GraphQlType.kt]: Any preference for this over `android.support.annotation.Nullable`?

File: apollo-compiler/src/main/kotlin/com/apollostack/compiler/ir/Field.kt (hunk 2 above)
Comment (felipecsl) [inline apollo-compiler/src/main/kotlin/com/apollostack/compiler/ir/Field.kt]: Yeah, actually the scalarTypes test uses a nested list and it seems to be working fine. I'll remove the TODO

File: apollo-compiler/src/main/kotlin/com/apollostack/compiler/ir/GraphQlType.kt [hunk 5]
```diff
@@ -1,11 +1,13 @@
 package com.apollostack.compiler.ir
 
+import com.apollostack.compiler.normalizeTypeName
+import com.squareup.javapoet.AnnotationSpec
 import com.squareup.javapoet.ClassName
 import com.squareup.javapoet.ParameterizedTypeName
 import com.squareup.javapoet.TypeName
+import javax.annotation.Nullable
```
Comment (felipecsl) [inline apollo-compiler/src/main/kotlin/com/apollostack/compiler/ir/GraphQlType.kt]: Either way works fine. I guess android support annotations is better. I'll change it

File: apollo-compiler/src/main/kotlin/com/apollostack/compiler/ir/GraphQlType.kt (hunk 4 above)
Comment (felipecsl) [inline apollo-compiler/src/main/kotlin/com/apollostack/compiler/ir/GraphQlType.kt]: will switch to android support

File: apollo-compiler/src/main/kotlin/com/apollostack/compiler/ir/GraphQlType.kt (hunk 1 above)
Comment (felipecsl) [inline apollo-compiler/src/main/kotlin/com/apollostack/compiler/ir/GraphQlType.kt]: good catch :)

**Comment by felipecsl** [comment]
As for the naming pattern for deduping, I propose we suffix it with `$`.
This is a fairly common practice, especially on annotation processors you see that a lot. This one is a slightly different scenario but I think it wouldn't sound that weird.
My second option would be to suffix with `_`. 
What do you guys think?

**Comment by martijnwalraven** [comment]
So if there are more than two duplicates this would become `$$` etc?

**Comment by felipecsl** [comment]
@martijnwalraven yeah, not pretty but gets the job done...

**Comment by sav007** [comment]
so lets stop on using '$' for now, to put this issue to bed :)

File: apollo-compiler/src/main/kotlin/com/apollostack/compiler/ir/GraphQlType.kt (hunk 5 above)
Comment (felipecsl) [inline apollo-compiler/src/main/kotlin/com/apollostack/compiler/ir/GraphQlType.kt]: android support annotations is neither on jcenter or maven central, so we're gonna have to stick to jsr305

**Comment by felipecsl** [comment]
Anyone has other comments on this? Otherwise merging on green build 🍏

**Comment by sav007** [comment]
LGTM

**Pull request merged**